create_openapi_definition.py

```diff
@@ -0,0 +1,213 @@
+# "paths".
+openapi_definition = {
+    "openapi": "3.0.1",
+    "info": {
+        "title": "Aportio ITSM REST API reference",
+        "description": "Reference documentation for Aportio's ITSM REST API.",
+        "version": "0.1"
+    },
+    "tags": [
+        {
+            "name": "Root",
+            "description": "The root level of the API"
+        },
+        {
+            "name": "Users",
+            "description": "Requests against user resources"
+        },
+        {
+            "name": "Tickets",
+            "description": "Requests against ticket resources"
+        },
+        {
+            "name": "Customers",
+            "description": "Requests against customer resources"
+        },
+        {
+            "name": "Comments",
+            "description": "Requests against comment resources"
+        },
+        {
+            "name": "Attachments",
+            "description": "Requests against attachment resources"
+        },
+        {
+            "name": "CustomerUserAssociations",
+            "description": "Requests against customer-user association resources"
+        }
+    ],
+    # It's nice to have the paths for our definitions in order, starting from Root, since they
+    # will show up in that order on Swaggerhub.
+    "paths": {path_value: dict() for key, path_value in RESOURCE_PATH_LOOKUP.items()}
+}
+
+# Create a YAML object from ruamel.yaml for loading and dumping yaml data.
+# Here we're using the ruamel.yaml library because it handles string block scalars better than
+# pyyaml.
+# A string block scalar in yaml is this:
+#
+# value: |
+#   {
+#       "_links": {
+#           "self": {
+#               "href": "/"
+#           },
+#           "users": {
+#               "href": "/users"
+#           },
+#           "customers": {
+#               "href": "/customers"
+#           },
+#           "tickets": {
+#               "href": "/tickets"
+#           },
+#           "comments": {
+#               "href": "/comments"
+#           },
+#           "attachments": {
+#               "href": "/attachments"
+#           },
+#           "customer_user_associations": {
+#               "href": "/customer_user_associations"
+#           }
+#       }
+#   }
+yaml = YAML()
+
+if __name__ == "__main__":
+    # Recursively go through the openapi_resource_definitions directory to get each resource's
+    # OpenAPI definition.
+    for dirpath, dirnames, files in os.walk("openapi_resource_definitions"):
+        # If dirnames is an empty list, it means we are at the end of a branch in the directory
+        # tree, which means the yaml files for one of the resources should be in this
+        # directory.
+        if not dirnames:
+            # Get the name of the resource that we're processing.
+            # dirpath will look something like:
+            #
+            # openapi_resource_definitions/individual_resources/Attachment
```

Comment:
'Attachment'? Is this specific, or just an example?

Reply:
Both, in a way. `for dirpath, dirnames, files in os.walk("openapi_resource_definitions")` will unpack tuples that look like this:

```
dirpath: openapi_resource_definitions/top_level_list_resources
dirnames: ['CustomerList', 'UserList', 'AttachmentList', 'CustomerUserAssociationList', 'CommentList', 'TicketList']
files: []

dirpath: openapi_resource_definitions/individual_resources/Attachment
dirnames: []
files: ['get.yaml']
```

Where `dirpath` is the path to the Attachment resource directory containing the `get.yaml` file. So that comment is an example of one of the possible directory paths

Reply:
I changed the comment slightly, I hope it makes it clearer

Reply:
Just a question, why are we giving plural in the tags alone? and singular everywhere else, is that how this is always done??

Reply:
Ah good point, I think I confused myself since I was adding multiple resources to a tag (e.g. Attachment, AttachmentList). I've changed the tags to be singular, since that's consistent with the idea of interacting with a single "resource"